In [560]:
import pandas as pd
from sklearn.model_selection import train_test_split
from pprint import pprint
import seaborn as sns

Loading in the data

In [561]:
X = pd.read_csv('./train.csv')
X_val = pd.read_csv('./test.csv')

Describing the X data

In [562]:
X.describe()

Id   MSSubClass  LotFrontage        LotArea  OverallQual  \
count  1460.000000  1460.000000  1201.000000    1460.000000  1460.000000   
mean    730.500000    56.897260    70.049958   10516.828082     6.099315   
std     421.610009    42.300571    24.284752    9981.264932     1.382997   
min       1.000000    20.000000    21.000000    1300.000000     1.000000   
25%     365.750000    20.000000    59.000000    7553.500000     5.000000   
50%     730.500000    50.000000    69.000000    9478.500000     6.000000   
75%    1095.250000    70.000000    80.000000   11601.500000     7.000000   
max    1460.000000   190.000000   313.000000  215245.000000    10.000000   

       OverallCond    YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1  ...  \
count  1460.000000  1460.000000   1460.000000  1452.000000  1460.000000  ...   
mean      5.575342  1971.267808   1984.865753   103.685262   443.639726  ...   
std       1.112799    30.202904     20.645407   181.066207   456.098091  ...   
min       1.000000  1872.000000   1950.000000     0.000000     0.000000  ...   
25%       5.000000  1954.000000   1967.000000     0.000000     0.000000  ...   
50%       5.000000  1973.000000   1994.000000     0.000000   383.500000  ...   
75%       6.000000  2000.000000   2004.000000   166.000000   712.250000  ...   
max       9.000000  2010.000000   2010.000000  1600.000000  5644.000000  ...   

        WoodDeckSF  OpenPorchSF  EnclosedPorch    3SsnPorch  ScreenPorch  \
count  1460.000000  1460.000000    1460.000000  1460.000000  1460.000000   
mean     94.244521    46.660274      21.954110     3.409589    15.060959   
std     125.338794    66.256028      61.119149    29.317331    55.757415   
min       0.000000     0.000000       0.000000     0.000000     0.000000   
25%       0.000000     0.000000       0.000000     0.000000     0.000000   
50%       0.000000    25.000000       0.000000     0.000000     0.000000   
75%     168.000000    68.000000       0.000000     0.000000     0.000000   
max     857.000000   547.000000     552.000000   508.000000   480.000000   

          PoolArea       MiscVal       MoSold       YrSold      SalePrice  
count  1460.000000   1460.000000  1460.000000  1460.000000    1460.000000  
mean      2.758904     43.489041     6.321918  2007.815753  180921.195890  
std      40.177307    496.123024     2.703626     1.328095   79442.502883  
min       0.000000      0.000000     1.000000  2006.000000   34900.000000  
25%       0.000000      0.000000     5.000000  2007.000000  129975.000000  
50%       0.000000      0.000000     6.000000  2008.000000  163000.000000  
75%       0.000000      0.000000     8.000000  2009.000000  214000.000000  
max     738.000000  15500.000000    12.000000  2010.000000  755000.000000  

[8 rows x 38 columns]

Describing the X testing data

In [563]:
X_val.describe()

Id   MSSubClass  LotFrontage       LotArea  OverallQual  \
count  1459.000000  1459.000000  1232.000000   1459.000000  1459.000000   
mean   2190.000000    57.378341    68.580357   9819.161069     6.078821   
std     421.321334    42.746880    22.376841   4955.517327     1.436812   
min    1461.000000    20.000000    21.000000   1470.000000     1.000000   
25%    1825.500000    20.000000    58.000000   7391.000000     5.000000   
50%    2190.000000    50.000000    67.000000   9399.000000     6.000000   
75%    2554.500000    70.000000    80.000000  11517.500000     7.000000   
max    2919.000000   190.000000   200.000000  56600.000000    10.000000   

       OverallCond    YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1  ...  \
count  1459.000000  1459.000000   1459.000000  1444.000000  1458.000000  ...   
mean      5.553804  1971.357779   1983.662783   100.709141   439.203704  ...   
std       1.113740    30.390071     21.130467   177.625900   455.268042  ...   
min       1.000000  1879.000000   1950.000000     0.000000     0.000000  ...   
25%       5.000000  1953.000000   1963.000000     0.000000     0.000000  ...   
50%       5.000000  1973.000000   1992.000000     0.000000   350.500000  ...   
75%       6.000000  2001.000000   2004.000000   164.000000   753.500000  ...   
max       9.000000  2010.000000   2010.000000  1290.000000  4010.000000  ...   

        GarageArea   WoodDeckSF  OpenPorchSF  EnclosedPorch    3SsnPorch  \
count  1458.000000  1459.000000  1459.000000    1459.000000  1459.000000   
mean    472.768861    93.174777    48.313914      24.243317     1.794380   
std     217.048611   127.744882    68.883364      67.227765    20.207842   
min       0.000000     0.000000     0.000000       0.000000     0.000000   
25%     318.000000     0.000000     0.000000       0.000000     0.000000   
50%     480.000000     0.000000    28.000000       0.000000     0.000000   
75%     576.000000   168.000000    72.000000       0.000000     0.000000   
max    1488.000000  1424.000000   742.000000    1012.000000   360.000000   

       ScreenPorch     PoolArea       MiscVal       MoSold       YrSold  
count  1459.000000  1459.000000   1459.000000  1459.000000  1459.000000  
mean     17.064428     1.744345     58.167923     6.104181  2007.769705  
std      56.609763    30.491646    630.806978     2.722432     1.301740  
min       0.000000     0.000000      0.000000     1.000000  2006.000000  
25%       0.000000     0.000000      0.000000     4.000000  2007.000000  
50%       0.000000     0.000000      0.000000     6.000000  2008.000000  
75%       0.000000     0.000000      0.000000     8.000000  2009.000000  
max     576.000000   800.000000  17000.000000    12.000000  2010.000000  

[8 rows x 37 columns]

Finding all NaN columns within the X data

In [564]:
nan_columns = [col for col in X.columns if X[col].hasnans]

nan_columns

['LotFrontage',
 'Alley',
 'MasVnrType',
 'MasVnrArea',
 'BsmtQual',
 'BsmtCond',
 'BsmtExposure',
 'BsmtFinType1',
 'BsmtFinType2',
 'Electrical',
 'FireplaceQu',
 'GarageType',
 'GarageYrBlt',
 'GarageFinish',
 'GarageQual',
 'GarageCond',
 'PoolQC',
 'Fence',
 'MiscFeature']

We want to approach the nan columns accordingly, measure the percentage of how many rows have NaN values before we start removing or imputing columns

In [565]:
def get_nan_percentage(column_name, dataset, total_rows):
    """
    Gets a %age of the NaN values within the column

    Args:
        column_name (_type_): The name of the column to analyze
        dataset (_type_): The dataset we are analyzing
        total_rows (_type_): The total # of rows the dataset has (used to calculate percentage)

    Returns:
        _type_: The percentage
    """
    nan_count = len(dataset[dataset[column_name].isna()])
    return round((nan_count / total_rows) * 100, 1)

In [566]:
percentages = {}
total_dataset_rows = X.shape[0]
for each_column in nan_columns:
    percentages[each_column] = '{}%'.format(get_nan_percentage(each_column, X, total_dataset_rows))

percentage_items = sorted(list(percentages.items()), key=lambda x: float(x[1].replace('%', '')))

pprint(percentage_items)

col_types = []

for each_item in percentage_items:
    col_types.append('{} = {}'.format(each_item[0], X[each_item[0]].dtype))

[('Electrical', '0.1%'),
 ('MasVnrType', '0.5%'),
 ('MasVnrArea', '0.5%'),
 ('BsmtQual', '2.5%'),
 ('BsmtCond', '2.5%'),
 ('BsmtFinType1', '2.5%'),
 ('BsmtExposure', '2.6%'),
 ('BsmtFinType2', '2.6%'),
 ('GarageType', '5.5%'),
 ('GarageYrBlt', '5.5%'),
 ('GarageFinish', '5.5%'),
 ('GarageQual', '5.5%'),
 ('GarageCond', '5.5%'),
 ('LotFrontage', '17.7%'),
 ('FireplaceQu', '47.3%'),
 ('Fence', '80.8%'),
 ('Alley', '93.8%'),
 ('MiscFeature', '96.3%'),
 ('PoolQC', '99.5%')]


Judging off this printout, there are columns that have barely any significance within the dataset, such as Alley, which has a 93.8% NaN count, and also PoolQC, which has a 99.5% rating, those ones can be dropped with little fallout.

There are 3 approaches to dealing with NaN columns: 1) Drop NaN columns, 2) Impute NaN columns with a constant or statistical method, 3) Extend imputation and perform a variant of One-Hot encoding, where we add another column that signifies the respective row was imputed, with either a 1 or 0.

In [567]:
from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import IterativeImputer, SimpleImputer
from sklearn.ensemble import RandomForestRegressor

nan_modified_x = X.copy()
nan_modified_x_val = X_val.copy()

#2 impute with statistical number

nan_modified_x['LotFrontage'].fillna(nan_modified_x['LotFrontage'].mean(), inplace=True)
nan_modified_x['MasVnrArea'].fillna(nan_modified_x['MasVnrArea'].mean(), inplace=True)
nan_modified_x_val['LotFrontage'].fillna(nan_modified_x_val['LotFrontage'].mean(), inplace=True)
nan_modified_x_val['MasVnrArea'].fillna(nan_modified_x_val['MasVnrArea'].mean(), inplace=True)

#1 drop NaN columns
nan_modified_x.drop(columns=['MiscFeature', 'PoolQC', 'Alley', 'Fence', 'FireplaceQu'], inplace=True)
nan_modified_x_val.drop(columns=['MiscFeature', 'PoolQC', 'Alley', 'Fence', 'FireplaceQu'], inplace=True)

# access the NaN columns left after dropping NaN columns
modified_nan_columns = [col for col in nan_modified_x if nan_modified_x[col].hasnans]
modified_nan_validation_columns = [col for col in nan_modified_x_val if nan_modified_x_val[col].hasnans]

# create simple imputer imputing by most frequent algorithm
imputer = SimpleImputer(strategy='most_frequent')

# Access imputed columns from imputing NaN columns
imputed_cols = pd.DataFrame(imputer.fit_transform(nan_modified_x[modified_nan_columns]))
validation_imputed_cols = pd.DataFrame(imputer.fit_transform(nan_modified_x_val[modified_nan_validation_columns]))

# Copying the index over
imputed_cols.index = nan_modified_x[modified_nan_columns].index
validation_imputed_cols.index = nan_modified_x_val[modified_nan_validation_columns].index

# Copying over the columns from the original to the imputed df
imputed_cols.columns = modified_nan_columns
validation_imputed_cols.columns = modified_nan_validation_columns

# Copying over the imputed column data to the original df
nan_modified_x[modified_nan_columns] = imputed_cols
nan_modified_x_val[modified_nan_validation_columns] = validation_imputed_cols

X = nan_modified_x.copy()
X_val = nan_modified_x_val.copy()

(X.shape, X_val.shape)

((1460, 76), (1459, 75))

Now that we've fixed all NaN values in the dataset, we can extract the target variable from the X dataset and the X_validation dataset

In [568]:
y = X['SalePrice'] # Extract the target variable from the dataset

X.drop(columns=['SalePrice'], inplace=True) # remove the column from the dataset

Now we attempt to verify the performance of all the categorical variable approaches, and how they affect the score of the model

In [569]:
from sklearn.metrics import mean_absolute_error
from sklearn.ensemble import RandomForestRegressor

def get_categorical_mae(x_train, x_test, y_train, y_test):
    """
    Gets the MAE of the random forest regressor

    Args:
        x_train (_type_): The features
        y_train (_type_): Target column
        x_test (_type_): Test feature dataset
        y_test (_type_): Test target column

    Returns:
        _type_: _description_
    """
    tree = RandomForestRegressor(random_state=0)
    tree.fit(x_train, y_train)
    predictions = tree.predict(x_test)
    return mean_absolute_error(predictions, y_test)

Approach 1: Drop all categorical columns without surgical decisions (basically meaning that we just drop them regardless of whether they are prominent in the dataset)

In [570]:
drop_c_X = X.copy() # copied X dataset
drop_c_X_validation = X_val.copy() # copy the validation dataset

In [571]:
categorical_columns = [x for x in drop_c_X.columns if drop_c_X[x].dtype == 'object']

drop_c_X.drop(columns=categorical_columns, inplace=True)

In [572]:
## split the data

cdrop_train_x, cdrop_val_x, cdrop_train_y, cdrop_val_y = train_test_split(drop_c_X, y, train_size=0.8, test_size=0.2, random_state=0)

get_categorical_mae(cdrop_train_x, cdrop_val_x, cdrop_train_y, cdrop_val_y)

18324.230239726025

Approach 2: Ordinal encoding

In [573]:
from sklearn.preprocessing import OrdinalEncoder

oe_x_df = X.copy()
oe_x_val_df = X_val.copy()
ordinal_encoder = OrdinalEncoder()

# Grabbing all categorical columns, which contain an column type of "object"
oe_categorical_columns = [col for col in oe_x_df.columns if oe_x_df[col].dtype == 'object']

# Create a new dataframe of the ordinal encoding
oe_encoded_categorical_columns = pd.DataFrame(ordinal_encoder.fit_transform(oe_x_df[oe_categorical_columns]))
oe_encoded_categorical_columns.index = oe_x_df.index # copy index over
oe_encoded_categorical_columns.columns = oe_categorical_columns # copy columns over
oe_x_df[oe_categorical_columns] = oe_encoded_categorical_columns # apply ordinal encoding to dataframe

# repeat same process with validation data
oe_encoded_val_categorical_columns = pd.DataFrame(ordinal_encoder.fit_transform(oe_x_val_df[oe_categorical_columns]))
oe_encoded_val_categorical_columns.index = oe_x_val_df.index
oe_encoded_val_categorical_columns.columns = oe_categorical_columns
oe_x_val_df[oe_categorical_columns] = oe_encoded_val_categorical_columns


In [574]:
oe_x_train, oe_x_val, oe_y_train, oe_y_val = train_test_split(oe_x_df, y) # splitting the encoded dataframe

In [575]:
get_categorical_mae(oe_x_train, oe_x_val, oe_y_train, oe_y_val)

20443.875534246577

Approach 3: One-Hot Encoding

In [576]:
from sklearn.preprocessing import OneHotEncoder

ohe_x = X.copy() # copy over dataframes
ohe_x_val = X_val.copy() # copy over dataframes
ohe = OneHotEncoder(sparse=False)

In [577]:
ohe_x_categorical_columns = [col for col in ohe_x.columns if ohe_x[col].dtype == 'object']
ohe_x_val_categorical_columns = [col for col in ohe_x_val.columns if ohe_x_val[col].dtype == 'object']

ohe_x_encoded = pd.DataFrame(ohe.fit_transform(ohe_x[ohe_x_categorical_columns]))
ohe_x_val_encoded = pd.DataFrame(ohe.fit_transform(ohe_x_val[ohe_x_val_categorical_columns]))


ohe_x_encoded.index = ohe_x.index # convert index over
ohe_x.drop(columns=ohe_x_categorical_columns, inplace=True) # drop columns to replace with OHE
ohe_x_val_encoded.index = ohe_x_val.index # convert index over
ohe_x_val.drop(columns=ohe_x_val_categorical_columns, inplace=True) # drop categorical columns to replace with OHE

ohe_x_concatted = pd.concat([ohe_x, ohe_x_encoded], axis=1)
ohe_x_val_concatted = pd.concat([ohe_x_val, ohe_x_val_encoded], axis=1)

ohe_x = ohe_x_concatted.copy()
ohe_x_val = ohe_x_val_concatted.copy()

In [578]:
ohe_x_train_, ohe_x_val_, ohe_y_train_, ohe_y_val_ = train_test_split(ohe_x, y)

In [579]:
get_categorical_mae(ohe_x_train_, ohe_x_val_, ohe_y_train_, ohe_y_val_)

c:\Users\flyin\anaconda3\lib\site-packages\sklearn\utils\validation.py:1688: FutureWarning: Feature names only support names that are all strings. Got feature names with dtypes: ['int', 'str']. An error will be raised in 1.2.
  warnings.warn(
c:\Users\flyin\anaconda3\lib\site-packages\sklearn\utils\validation.py:1688: FutureWarning: Feature names only support names that are all strings. Got feature names with dtypes: ['int', 'str']. An error will be raised in 1.2.
  warnings.warn(


15437.908191780822

In [581]:
from sklearn.metrics import mean_absolute_error
final_model = RandomForestRegressor(n_estimators=155)
final_model.fit(oe_x_df, y)
final_model_predictions = final_model.predict(oe_x_val_df)

resultant_df = pd.DataFrame({ 'Id': oe_x_val_df.Id, 'SalePrice': final_model_predictions })

resultant_df.to_csv('./submission2.csv', index=False)